baggage_restriction_manager.go

```diff
@@ -0,0 +1,153 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package jaeger
+
+import (
+	"fmt"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/uber/jaeger-client-go/thrift-gen/baggage"
+	"github.com/uber/jaeger-client-go/utils"
+)
+
+const (
+	defaultMaxValueLength = 2048
+)
+
+// BaggageRestrictionManager keeps track of valid baggage keys and their size restrictions.
+type BaggageRestrictionManager interface {
+	// IsValidBaggageKey returns whether the baggage key is valid given the restrictions
+	// and the max baggage value length
+	IsValidBaggageKey(key string) (bool, int32)
+}
+
+// DefaultBaggageRestrictionManager allows any baggage key.
+type DefaultBaggageRestrictionManager struct{}
+
+// IsValidBaggageKey implements BaggageRestrictionManager#IsValidBaggageKey
+func (m DefaultBaggageRestrictionManager) IsValidBaggageKey(key string) (bool, int32) {
+	return true, defaultMaxValueLength
+}
+
+type httpBaggageRestrictionsManager struct {
+	serverURL string
+}
+
+func (s *httpBaggageRestrictionsManager) GetBaggageRestrictions(serviceName string) ([]*baggage.BaggageRestriction, error) {
+	var out []*baggage.BaggageRestriction
+	v := url.Values{}
+	v.Set("service", serviceName)
+	if err := utils.GetJSON(s.serverURL+"?"+v.Encode(), &out); err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+type baggageRestrictionManager struct {
+	baggageRestrictionManagerOptions
+	sync.RWMutex
+
+	serviceName     string
+	restrictionsMap map[string]int32
+	timer           *time.Ticker
+	thriftManager   baggage.BaggageRestrictionManager
+	pollStopped     sync.WaitGroup
+	stopPoll        chan struct{}
+}
+
+// NewBaggageRestrictionManager returns a BaggageRestrictionManager that polls the agent for the latest
+// baggage restrictions.
+func NewBaggageRestrictionManager(
+	serviceName string,
+	options ...BaggageRestrictionManagerOption,
+) (BaggageRestrictionManager, error) {
+	opts := applyOptions(options...)
+	m := &baggageRestrictionManager{
+		serviceName:                      serviceName,
```

Comment:
there is a developing use case, e.g. with Envoy or Muttley, where a single tracer can generate traces on behalf of many services. Perhaps we don't need to store the service name here, since it's being passed as an argument to the Get method anyway?